### PR TITLE
Added update command and rewritten install command

### PR DESCRIPTION
**Pull request by chris-steele**
### Changes in this PR:

New command `update`
 - resolves https://github.com/adaptlearning/adapt-cli/issues/61
 - allows users to update/change installed plugins
 - switch `--check` allows user to see what updates are available
 - user is warned if attempting to use incompatible or out of date plugins and prompted to choose course of action
 - provides progress info and summary

Command `install` rewritten
 - new switch `--compatible` ensures only the latest compatible plugins are installed
 - new switch `--dry-run` allows user to see what the command will do without making any changes
 - user is warned if attempting to install incompatible or out of date plugins and prompted to choose course of action
 - provides progress info and summary

### Rationale for `install` rewrite

If the master version of manifest (`adapt.json`) is kept up-to-date then appropriate versions will be installed when `adapt install` is run in a new project. When installing an older version of the framework (for testing purposes, for legacy plugin/browser support etc) the manifest may not be appropriate and so incompatible plugins may be installed. Manually determining appropriate versions of plugins for a project is tedious and error prone (and equally for the master version).

With these changes the manifest is still respected; so if a plugin is constrained to a semver then the command will attempt to install an applicable version. If '*' is specified then the latest compatible version is installed.

A consequence of these changes is that it will be no longer necessary to amend the master version of the manifest and we can return to using the wildcard '*'.

### Notes

Install the prerelease that contains these changes by doing `npm install -g adapt-cli@next`

### CODE REVIEW

**Comment by chris-steele** [comment]
@moloko, @tomgreenfield install command now runs with untagged plugins. Is there anything else needed before requesting +1s?

**Comment by moloko** [comment]
@chris-steele just a bit of an indentation sort out if you don't mind? currently a mix of tabs & spaces

**Comment by chris-steele** [comment]
@moloko checked update and install scripts - they're all using space indentation as far as I can see. Did you mean globally rather than these two scripts?

**Comment by moloko** [comment]
as far as I can see all the files in this PR are using a mix of tabs and spaces - sometimes in the same line! I'll mark a few with comments so you can see what I mean.

File: lib/commands/install/InstallLog.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
+var readline = require('readline');
+
+var isLoggingProgress = false;
+var lastProgressStr = '';
+
+module.exports = {
+	logProgress:function(str) {
+		lastProgressStr = str;
+		isLoggingProgress = true;
+		readline.cursorTo(process.stdout, 0);
```
Comment (moloko) [inline lib/commands/install/InstallLog.js]: tab indentation

File: lib/commands/install/InstallLog.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        process.stdout.write(str);
```
Comment (moloko) [inline lib/commands/install/InstallLog.js]: space indentation

File: lib/commands/install/InstallTarget.js [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+	constructor: function(name, requestedVersion, isCompatibleEnabled) {
+        var endpoint = name+'#'+(isCompatibleEnabled ? any : requestedVersion),
+			ep = endpointParser.decompose(endpoint),
+	        version = /^\*$/.test(ep.target) ? any : ep.target,
+	        proposal;
+
+	    Plugin.call(this, ep.name || ep.source, version);
+
+	    if (this.version !== any) {
```
Comment (moloko) [inline lib/commands/install/InstallTarget.js]: tabs AND spaces!

Comment (chris-steele) [reply lib/commands/install/InstallTarget.js]: sorry I was literally just looking at the big two files in this PR (install and update) and forgot about the dependencies. Is there a preference - tabs or spaces?

File: lib/commands/install/InstallLog.js (hunk 2 above)
Comment (chris-steele) [reply lib/commands/install/InstallLog.js]: @moloko hopefully caught all the files I've edited in this PR

**Pull request merged**